### PR TITLE
Reduce locking on IssueCredit path

### PR DESCRIPTION
**Pull request by danielmarbach**
TODO

### CODE REVIEW

File: src/AmqpLink.cs [hunk 1]
```diff
@@ -1284,16 +1281,13 @@ void SendFlowInternal(bool echo, bool drain, Fields properties)
             do
             {
                 Flow flow = new Flow();
-                flow.Handle = this.LocalHandle;
-                lock (this.syncRoot)
+                flow.Handle = this.settings.Handle;
```
Comment (danielmarbach) [inline src/AmqpLink.cs:1284]: The assumption here was that all those values are 32 bit and can be read atomically

File: src/AmqpSession.cs [hunk 2]
```diff
@@ -1115,13 +1111,10 @@ public bool TrySendTransfer(Delivery delivery, Transfer transfer, ByteBuffer pay
...
-                {
-                    // Outgoing state needs to be sync'ed with transfers
-                    this.AddFlowState(flow, false);
-                    this.Session.incomingChannel.AddFlowState(flow, true);
-                    this.Session.SendCommand(flow, null);
-                }
+                // Outgoing state needs to be sync'ed with transfers
+                this.AddFlowState(flow);
+                this.Session.incomingChannel.AddFlowState(flow, true);
+                this.Session.SendCommand(flow, null);
```
Comment (danielmarbach) [inline src/AmqpSession.cs:1117]: Command sending seems to be thread safe

File: src/AmqpSession.cs [hunk 3]
```diff
@@ -1168,13 +1161,10 @@ public void OnFlow(Flow flow)
...
 
-            public void AddFlowState(Flow flow, bool reset)
+            public void AddFlowState(Flow flow)
             {
-                lock (this.SyncRoot)
-                {
-                    flow.OutgoingWindow = this.outgoingWindow;
-                    flow.NextOutgoingId = this.nextOutgoingId.Value;
-                }
+                flow.OutgoingWindow = this.outgoingWindow;
```
Comment (danielmarbach) [inline src/AmqpSession.cs:1166]: Reading 32 bit values

**Comment by danielmarbach** [comment]
Closing for now since the field access patterns used in these classes truly requires atomic modifications and I cannot say I fully understand it yet.

**Comment by xinchen10** [comment]
Really appreciate your work to improve the library. AMQP semantically is single threaded at the connection level. Ordering is critical to maintain protocol state. This is the main reason the current implementation uses an object level lock to protect protocol state and any operations that could change that state (e.g. SendFlow). Some places could use a smaller coped lock (e.g. message queue and waiters in receiving link) but that does not change much because the lock contention just goes to a different place where things need to be synchronized. Without locking the commands would go on wire out of order. For example, the app issues credit 1, 3, 5 in that order and the flow to the peer could be 1, 5, 3. 

This implementation is optimized for the message transfer scenarios, where you set up a link with some reasonable credits and handle messages from the OnMessage handler. The receive APIs have too much overhead, especially when they are called concurrently.

Long term solution to this problem would be to maintain a queue for operations that change protocol state and a single thread to process the work from that queue. The queue could be at the session level or the connection level. Links require delivery ID and state that are managed by session so we cannot implement the operation queue at the link level efficiently.

Right now there are some optimizations we could do. For example, IssueCredit could be called out side of the lock if we add extra logic to maintain calling order.

**Comment by danielmarbach** [comment]
I've done a few local spikes to replace for example the buffer management with ArrayPool which is doable but then I stumbled over some edge cases I wasn't sure about and stopped.

I think so many of the synchronisation here could be replaced with System.Threading.Channel allowing multiple writers but a single reader. Maybe I'll find some time. Would you be interested to see those spikes here as a means of seeing the thoughts which might trigger other ideas?

**Pull request closed without merging**